src/dataflow-types/src/reconciliation/command.rs

```diff
@@ -0,0 +1,361 @@
+        if let Some(response) = self.responses.pop_front() {
+            Some(response)
+        } else {
+            let response = self.client.recv().await;
+            if let Some(response) = response {
+                self.absorb_response(response)
+            }
+            self.responses.pop_front()
+        }
+    }
+}
+
+impl<C: Client> CommandReconcile<mz_repr::Timestamp, C> {
+    /// Construct a new [CommandReconcile].
+    ///
+    /// * `client`: The client wrapped by this struct.
+    pub fn new(client: C) -> Self {
+        Self {
+            client,
+            created: false,
+            responses: Default::default(),
+            storage_initialized: false,
+            uppers: Default::default(),
+        }
+    }
+
+    /// Start tracking of a id within an instance.
+    ///
+    /// If we're already tracking this ID, it means that the coordinator lost connection and
+    /// reconnected (or has a bug). We're updating the coordinator's upper to match the local state.
+    fn start_tracking(&mut self, id: GlobalId, instance: Option<ComputeInstanceId>) {
+        let frontier = timely::progress::frontier::MutableAntichain::new_bottom(
+            <mz_repr::Timestamp as timely::progress::Timestamp>::minimum(),
+        );
+        match self.uppers.entry((id, instance)) {
+            Entry::Occupied(entry) => {
+                let mut change_batch = ChangeBatch::new_from(
+                    <mz_repr::Timestamp as timely::progress::Timestamp>::minimum(),
+                    -1,
+                );
+                change_batch.extend(entry.get().frontier().iter().copied().map(|t| (t, 1)));
+                // dbg!(entry.get());
+                // dbg!(&change_batch);
+                if let Some(instance) = instance {
+                    self.responses.push_back(Response::Compute(
+                        ComputeResponse::FrontierUppers(vec![(id, change_batch)]),
+                        instance,
+                    ));
+                } else {
+                    let feedback = TimestampBindingFeedback {
+                        changes: vec![(id, change_batch)],
+                        bindings: vec![],
+                    };
+                    self.responses.push_back(Response::Storage(
+                        StorageResponse::TimestampBindings(feedback),
+                    ));
+                }
+            }
+            Entry::Vacant(entry) => {
+                entry.insert(frontier);
+            }
+        }
+    }
+
+    /// Stop tracking the id within the instance.
+    fn stop_tracking(&mut self, id: GlobalId, instance: Option<ComputeInstanceId>) {
+        let previous = self.uppers.remove(&(id, instance));
+        if previous.is_none() {
+            warn!("Protocol error: ceasing frontier tracking for absent identifier {id:?}");
+        }
+    }
+
+    async fn observe_command(&mut self, command: Command) -> Result<(), anyhow::Error> {
```

Comment:
Minor nit, but elsewhere (in `client::partitioned` I think?) the name `observe` gets applied to a `&Command`, and the meaning is "do the things you need to do in response, but someone else will send the command". It might be different here, and if so we probably want either a different name or some comments!

Reply:
Changed to `absorb_command` to stay in line with `absorb_response`.